pyomo/solvers/plugins/solvers/xpress_direct.py

```diff
@@ -223,8 +223,268 @@ def _apply_solver(self):
+        elif status == xp.mip_infeas: # MIP proven infeasible
+            results.solver.status = SolverStatus.warning
+            results.solver.termination_message = "Model was proven to be infeasible"
+            results.solver.termination_condition = TerminationCondition.infeasible
+            soln.status = SolutionStatus.infeasible
+        elif status == xp.mip_optimal: # optimal
+            results.solver.status = SolverStatus.ok
+            results.solver.termination_message = "Model was solved to optimality (subject to tolerances), " \
+                "and an optimal solution is available."
+            results.solver.termination_condition = TerminationCondition.optimal
+            soln.status = SolutionStatus.optimal
+        elif status == xp.mip_unbounded and mip_sols > 0:
+            results.solver.status = SolverStatus.warning
+            results.solver.termination_message = "LP relaxation was proven to be unbounded, " \
+                "but a solution is available."
+            results.solver.termination_condition = TerminationCondition.unbounded
+            soln.status = SolutionStatus.unbounded
+        elif status == xp.mip_unbounded and mip_sols <= 0:
+            results.solver.status = SolverStatus.warning
+            results.solver.termination_message = "LP relaxation was proven to be unbounded."
+            results.solver.termination_condition = TerminationCondition.unbounded
+            soln.status = SolutionStatus.unbounded
+        else:
+            results.solver.status = SolverStatus.error
+            results.solver.termination_message = \
+                ("Unhandled Xpress solve status "
+                 "("+str(status)+")")
+            results.solver.termination_condition = TerminationCondition.error
+            soln.status = SolutionStatus.error
+
+        results.problem.upper_bound = None
+        results.problem.lower_bound = None
+        if xprob_attrs.objsense == 1.0:  # minimizing MIP
+            try:
+                results.problem.upper_bound = xprob_attrs.mipbestobjval
+            except (XpressDirect.XpressException, AttributeError):
+                pass
+            try:
+                results.problem.lower_bound = xprob_attrs.bestbound
+            except (XpressDirect.XpressException, AttributeError):
+                pass
+        elif xprob_attrs.objsense == -1.0:  # maximizing MIP
+            try:
+                results.problem.upper_bound = xprob_attrs.bestbound
+            except (XpressDirect.XpressException, AttributeError):
+                pass
+            try:
+                results.problem.lower_bound = xprob_attrs.mipbestobjval
+            except (XpressDirect.XpressException, AttributeError):
+                pass
+            
+        # TODO: Pyomo only queries solution if the LP status is optimal.
+        #       This seems wrong at first glance
+        return (xprob_attrs.lpstatus == xp.lp_optimal and mip_sols > 0,
+                mip_sols > 0)
```

Comment:
I agree with the `TODO` (and it was almost certainly my mistake originally). I think the `_get_*_results` functions should just return a single flag, "is a solution available or not". This will simplify things in several places.